src/dependency-checks.js

```diff
@@ -0,0 +1,265 @@
+  let addons = {};
+
+  for (let i = 0; i < REQUIRED_DATA_PACKAGES.length; i++) {
+    let requiredPackageName = REQUIRED_DATA_PACKAGES[i];
+
+    let pkgInfos = (addons[requiredPackageName] = checkForPackage(
+      checker,
+      requiredPackageName,
+      m3Versions,
+      dataInfo
+    ));
+
+    if (hasFullEmberData) {
+      // we have a required package that doesn't match the ember-data version
+      if (
+        pkgInfos.uniqueDeps.length === 1 &&
+        requiredPackageName !== 'ember-inflector' &&
+        pkgInfos.uniqueDeps[0].pkg.version !== fullEmberDataVersion
+      ) {
+        throw new Error(
+          `ember-m3 "${pkg.version}" requires the peerDependency ${requiredPackageName} to match the installed ember-data version "${fullEmberDataVersion}". Found ${requiredPackageName} "${pkgInfos.uniqueDeps[0].pkg.version}".`
+        );
+      }
+      // we don't have a required package at all
+    } else if (pkgInfos.uniqueDeps.length !== 1) {
+      throw new Error(
+        `ember-m3 "${pkg.version}" requires the peerDependency ${requiredPackageName} to be installed.`
+      );
+      // we have a required package, check min version
+    } else {
+      let pkgVersion = pkgInfos.uniqueDeps[0].pkg.version;
+      if (
+        semver.lt(pkgVersion, minVersionForUsingOwnPackages) &&
+        requiredPackageName !== 'ember-inflector'
+      ) {
+        throw new Error(
+          `To use your own @ember-data/<pkg> versions with ember-m3 their versions must be at least "${minVersionForUsingOwnPackages}". Found ${requiredPackageName} with version "${pkgVersion}".`
+        );
+      }
+    }
+  }
+
+  let storeVersion = fullEmberDataVersion || addons['@ember-data/store'].uniqueDeps[0].pkg.version;
+
+  for (let i = 0; i < ALL_DATA_PACKAGES.length; i++) {
+    let addonName = ALL_DATA_PACKAGES[i];
+    let pkgInfos = (addons[addonName] =
+      addons[addonName] || checkForPackage(checker, addonName, m3Versions, dataInfo));
+    hasSingleVersionIfPresent(pkgInfos);
+
+    if (pkgInfos) {
+      let addon = pkgInfos.uniqueDeps.length ? pkgInfos.uniqueDeps[0] : null;
+      let pkgVersion = addon && addon.pkg.version;
+      if (addon && pkgVersion !== storeVersion) {
+        throw new Error(
+          `All @ember-data/<pkg> packages must have matching versions. Found ${addonName} "${pkgVersion}" which does not match @ember-data/store's version of "${storeVersion}".`
+        );
+      }
+    }
+  }
+
+  hasSingleVersionIfPresent(addons['ember-inflector']);
+}
+
+function getOwnVersions() {
+  let versions = {};
+  M3DataDeps.forEach(name => {
+    versions[name] = require(`${name}/package.json`).version;
+  });
+
+  return versions;
+}
+
+/**
+ * Throws errors if invalid ember-data or ember-m3 configurations are detected.
+ * Returns a boolean `true` if the consuming app brings it's own versions of
+ * ember-data or ember-data packages. `false` otherwise.
```

Comment:
The comment doens't mention ember-inflector but at least some checks here check for it. Great comment otherwise though 👍

Reply:
Updated